### PR TITLE
cmd: Implement storage import/export

### PR DESCRIPTION
**Pull request by Phrynobatrachus**
#5313 

Thanks for the fun project!

1)
> I think I'd probably want to put that in the caddycmd package though (the cmd folder), as it's purely a CLI utility, and not really anything to do with Caddy core. (Assuming nothing new needs to be exported. If so let me know.)

The two functions in `caddy.go` that is? There are caddycmd wrappers separately as is.

2)
> In the case of running as a systemd service, the command would need to be run as the caddy user to correctly access the files owned by that user, if the file_system storage module is used.

I started by shelling out `systemctl status caddy` for part of the rerun check but relying only on the `--user` flag felt a little less awkward. I'm not sure what the best option is though. Besides file permissions, it seemed necessary to re-exec with the corresponding alternate user' `$HOME` set, for AppDataDir() -> homeDirUnsafe(), so the user has to be specified anyways to get that.

- Windows handling (AppData vs HOME), etc?

- Add `CAP_SETUID` / `CAP_SETGID` to `setcap.sh`?

- fork/exec error on the user switch without `NoSetGroups: true` https://walac.github.io/golang-patch/

3)
`run()` vs `Run()`

```
// This is a low-level function; most callers
// will want to use Run instead, which also
// updates the config's raw state.
```
run() makes sense to me for exporting, maybe importing should Run() instead?

### CODE REVIEW

File: cmd/commands.go
Comment (francislavoie) [reply cmd/commands.go:309]: :thinking: yeah either one is fine. I thought it would be fun to set up a subcommand. I might take a look at it in the coming days, I'm curious to see how it would work.

File: caddy.go
Comment (mholt) [inline caddy.go:680]: Rather than running the config, maybe better to load and unmarshal just the `storage` portion. You could try doing something like this:

https://github.com/caddyserver/caddy/blob/cfc85ae8caf032c3ef92675a024fc5d1264ab7ce/cmd/commandfuncs.go#L693-L717

You might need to make a caddy.Context and then call LoadModule() on it though -- I haven't tried doing that in isolation like this but I'd be happy to help if that is problematic...

File: caddy.go
Comment (mholt) [inline caddy.go:676]: I'm hoping we can find a way to put these functions -- unexported -- in the `caddycmd` package.

File: caddy.go
Comment (mholt) [inline caddy.go:678]: Loading the entire storage into memory is probably not ideal -- it will blow up for large deployments (where a feature like this is most useful).

Maybe exporting should write directly [to an archive](https://github.com/mholt/archiver).

Although, we're getting into separate tooling territory.

What if, instead, we had a `transfer storage` utility that simply moved the contents of one storage to another storage? We could do it kv-pair by kv-pair, much more efficiently, directly, without any need to read everything into memory or have separate import/export ops.

**Review by mholt — CHANGES_REQUESTED** [review]
Thanks for this implementation!

While reviewing I thought of something that might make more practical sense... basically a `transfer-storage` utility that does both export and import at the same time, in lock-step, to avoid blowing up memory and also for simplicity's sake.

We'd just need to figure out how to configure both storage modules.

File: caddy.go
Comment (Phrynobatrachus) [reply caddy.go:680]: Yeah I did try it this way to easily get at a caddy.Context, I'll give that a shot.

File: caddy.go
Comment (Phrynobatrachus) [reply caddy.go:678]: > What if, instead, we had a transfer storage utility that simply moved the contents of one storage to another storage? We could do it kv-pair by kv-pair, much more efficiently, directly, without any need to read everything into memory or have separate import/export ops.

Yeah that's a good point, like just Load() from old -> Store() in new -> Delete() from old per each key at a time.

Comment (mholt) [reply caddy.go:678]: Yes! Errrr, but maybe make Delete() opt-in, and maybe that should only be done after the migration is 100% totally completed successfully.

Comment (francislavoie) [reply caddy.go:678]: Transfer alone might not be enough in situations where you want to move from one cluster to another where the storage is completely isolated in the old vs new. I think adding transfer is a good idea, but I don't think it should be instead of export/import, I think both should exist.

Comment (mholt) [reply caddy.go:678]: I don't think backups are in the scope of the web server. I'd like to leave that up to backup utilities.

I also would like to hear of actual use cases that require an export but not an import -- AFAIK the need is to be able to switch storage providers, migrate certs from one place to another. A plain export can be done with the storage system itself (i.e. copy the files from disk, or run a SQL query) -- importing so that Caddy can use it is harder, and it makes sense to be able to import from any data source.

The more I think about this the more I'm convinced that we need a transfer utility -- effectively an "import", but of course supporting any storage module as a source -- not an exporter.

Comment (francislavoie) [reply caddy.go:678]: Like I said, if the storage is isolated in such a way that Caddy cannot contact both locations, then transfers cannot work. I don't think we can presume that transfers are always possible. That would be a very unfortunate limitation.

And for backups, that assumes the users know how to perform backups using their storage mechanism, and that they might even care to backup anything other than Caddy's data in that storage. For example for Redis, I might be using it as a cache otherwise, so I don't care to backup anything else.

Also re memory usage of loading everything in storage, I don't think that's a relevant concern. The CLI script is what will be keeping it in memory. Each LE cert+key+json is roughly 6k in size (with default key type etc), so 1000 certs (which is quite large) is ~6MB, and a huge deployment of like 50,000 certs would be like ~300MB; on a deployment that large, they'd surely be using a machine with plenty of RAM, hopefully over-provisioned.

Either way, writing straight to a mutable archive file is a good idea but I wouldn't consider that absolutely necessary.

Comment (Phrynobatrachus) [reply caddy.go:678]: > Also re memory usage of loading everything in storage, I don't think that's a relevant concern.

fwiw this was my first impression as well -- that everything in storage is relatively small both individually and overall, and that the storage is strictly an interface for items critical to operation rather than anything served by itself ("must not be a cache..." note in the docs). I'm definitely not experienced with any large deployments though.

Comment (Phrynobatrachus) [reply caddy.go:678]: > A plain export can be done with the storage system itself (i.e. copy the files from disk, or run a SQL query) -- importing so that Caddy can use it is harder, and it makes sense to be able to import from any data source.

I agree that exporting already has some clear ways to do it. The framing that makes sense to me is that commands as originally described seem useful to help _facilitate_ transfers/backups (either manually or scripted), but actually doing those would indeed be out of scope. So it's a question of whether it's too little to justify adding rather than too much. 

> The more I think about this the more I'm convinced that we need a transfer utility -- effectively an "import", but of course supporting any storage module as a source -- not an exporter.

I don't follow quite as much with this. I don't want to muddle things too much, I guess my thought for now is any potential memory issue seems orthogonal to whatever CLI semantics are desired.

Comment (mholt) [reply caddy.go:678]: I feel like it's just bad form though -- even if the data happens to only be a few MB in most cases, we do know there are clusters out there with 100,000+ certs. It's so easy to stream the contents file-by-file, so we should probably do that. If someone comes to us complaining that there's a lot of memory use when doing an import/export (because, memory is a shared resource), then I would have no good excuse for that.

Comment (francislavoie) [reply caddy.go:678]: Which is why I'm saying it makes sense to have both import/export _and_ transfer. It's up to the user to pick which is best for them. But only having transport and not import/export isn't acceptable IMO.

File: cmd/commands.go
Comment (francislavoie) [reply cmd/commands.go]: Pushed a commit, got it working :+1: was pretty simple.

Also as an aside, please make sure your editor is configured to use `.editorconfig`; your code was using spaces instead of tabs among other minor style issues. Use `gofmt -w -s <path>` to apply fixes.

**Comment by mholt** [comment]
If we aren't going to make this a transfer utility, I'll revisit this after the 2.7 beta 1 release, which is my current priority. I'm still not sure exporting data from storage systems is within the scope of the web server.

**Comment by francislavoie** [comment]
But only Caddy can provide a unified interface to export/import for any storage implementation, because of the interfaces. I think it's definitely in scope, when `storage` is a top-level configuration option in Caddy.

But yeah, no rush on this, it's not a blocker for 2.7.0.

File: cmd/commandfuncs.go
Comment (francislavoie) [reply cmd/commandfuncs.go]: Rethinking this, I'm not sure we need `--user` at all. I think we should just add a mention that the command should be run as the same user Caddy runs as (typically `caddy` on Linux with systemd). Linux users can use `sudo su caddy -c 'caddy export -o <wherever>` to run it as the correct user.

On Windows, I think users typically end up running it as their current administrator user, so it shouldn't be a concern.

As you can see from the Windows CI build, it's failing because of the use of `syscall.Credential` which isn't defined on Windows. So I think we should avoid trying to muck with retrying as a different user.

Thanks for trying it though, it's an interesting experiment either way.

File: cmd/commandfuncs.go
Comment (francislavoie) [reply cmd/commandfuncs.go:551]: I think what might be more useful is to write an "are you sure" message showing what the current user is and the current storage path if the default storage is configured, to give users a chance to realize their mistake if they ran it incorrectly.

As an aside, I noticed that running the command without `-o` causes the tar contents to just get dumped to stdout. I don't think we should do that by default. I think we should require `-o` but allow using `-` as a value to mean write to stdout. That way users can do `caddy storage export -o- | some-other-command-that-takes-it-as-input`.

On that topic, I think users could do this to perform a transfer in a one-liner without writing to a file:

```
caddy storage export --config Caddyfile.old -o- | caddy storage import --config Caddyfile.new -i-
```

File: cmd/commandfuncs.go
Comment (Phrynobatrachus) [reply cmd/commandfuncs.go]: Cool. I saw the CI failure and started looking at the Windows syscall interface (https://pkg.go.dev/golang.org/x/sys/windows#CreateProcessAsUser) but it's certainly simpler without all of that in the first place.

**Comment by mholt** [comment]
An export without an import is just a backup, and that's not Caddy's job. If you want to back up your certs, run `rsync` on your Caddy data dir, or a SQL dump of your storage DB, or do the equivalent for whatever storage medium you're using. Then you can restore later using that same tooling, no need for Caddy to reinvent the wheel.

Now, an import is definitely useful. But import without source data is useless, so really what we need is an _import_ feature that can get the contents from a configured storage, which so happens to basically be a kind of export.

So, the user can still do what they need (export now, import later), it's just that they don't use Caddy until that import step.

**Comment by Phrynobatrachus** [comment]
I pushed a commit that gets rid of the `caddy.go` import/export functions and loads the storage modules solely from the caddycmd package. Otherwise left the CLI setup the same for the moment. Hopefully a little closer though!

File: cmd/commandfuncs.go
Comment (Phrynobatrachus) [inline cmd/commandfuncs.go:554]: This doesn't correctly handle the default case where no storage is defined in the config. My current thought is to check `json.Valid` first so that if `Unmarshal` returns `json.SyntaxError` it should be safe to assume it's the default case.

Comment (mholt) [reply cmd/commandfuncs.go:554]: That could probably work.

**Review by mholt — CHANGES_REQUESTED** [review]
Getting there!

I still need to decide whether to have this as two commands (import/export) or just one (transfer).

**Comment by mholt** [comment]
@Phrynobatrachus Thanks so much for continuing to work on this and for your patience. It's a high-quality contribution -- I am just dragging my feet on deciding what to do here.

We'll get it in sooner or later though :sweat_smile:

**Comment by Phrynobatrachus** [comment]
Take your time! I'm here for the practice anyways, it's all good.

**Comment by Phrynobatrachus** [comment]
Some notes from testing against storage implementations with the transfer invocation suggested by Francis:

`caddy storage export -c Caddyfile.old -o- | caddy storage import -c Caddyfile.new -i-`

(where old is file_system, new is alt storage)

## postgres-storage

- doesn't support recursive parameter in `List()`, which this PR uses

https://github.com/yroc92/postgres-storage/blob/276797aefe401b738781692d278a158c53b99208/storage.go#L317C2-L319

```go
      if recursive {
              return nil, fmt.Errorf("recursive not supported")
      }
```
## caddy-tlsredis

- works after a patch to `Stat()`

https://github.com/gamalan/caddy-tlsredis/blob/66e1c61412decebb977790af25f7a6ce5c9b5878/storageredis.go#L488-L493

```go
 return certmagic.KeyInfo{
     Key:        key,
     Modified:   data.Modified,
     Size:       int64(len(data.Value)),
     IsTerminal: false,   
}, nil
```
changed to `IsTerminal: true`, appropriate for all cases? 

## certmagic-s3

- works after a patch to Stat()

https://github.com/ss098/certmagic-s3/blob/627ff2e5b5a473e29cc63f15e37cbd6b1b22eeab/s3.go#L244-L249

```go
        return certmagic.KeyInfo{
                Key:        object.Key,
                Modified:   object.LastModified,
                Size:       object.Size,
                IsTerminal: strings.HasSuffix(object.Key, "/"),
        }, err
```

should be `IsTerminal: !strings.HasSuffix(object.Key, "/")`

**Comment by francislavoie** [comment]
Ooh very interesting, thanks for trying that!

I don't know the answer re `IsTerminal`, Matt will need to answer that.

**Comment by mholt** [comment]
@Phrynobatrachus Excellent, thanks for the field trials.

> changed to IsTerminal: true, appropriate for all cases?

Only when the stat'ed key is a "file" (i.e. not a "directory"). In other words, only when the key doesn't contain other keys within/under it. If that makes sense.

> should be IsTerminal: !strings.HasSuffix(object.Key, "/")

Probably true! Good catch.

> doesn't support recursive parameter in List(), which this PR uses

@yroc92 Can I help with discussing how to support `recursive=true` on postgres-storage?

**Comment by francislavoie** [comment]
If export always essentially turns whatever storage into the layout it would have if it was the filesystem module, then I think that's pretty natural. It's not like we're introducing a new format of somekind, it's already how storage already exists, just put in a tarball.

IMO the only downside of import/export is that it's loaded in memory before writing out. But I think that's fine for a large majority of users. If a truly huge storage needs to be migrated then we can introduce `transfer`. But IMO import+export is the minimum viable product here.

**Comment by Phrynobatrachus** [comment]
> Only when the stat'ed key is a "file" (i.e. not a "directory"). In other words, only when the key doesn't contain other keys within/under it. If that makes sense.

Yeah, if `!strings.HasSuffix(object.Key, "/")` is a reasonable expression of that for certmagic-s3, caddy-tlsredis can probably just do the same.

**Comment by mholt** [comment]
> Yeah, if !strings.HasSuffix(object.Key, "/") is a reasonable expression of that for certmagic-s3, caddy-tlsredis can probably just do the same.

I _think_ that depends on how the storage mechanism is implemented. For example, file systems don't necessarily require a trailing `/` to indicate a directory -- whether it's a directory is stored as metadata.

**Comment by Phrynobatrachus** [comment]
> IMO the only downside of import/export is that it's loaded in memory before writing out.

This was only a problem with my original version where I was unnecessarily buffering the data to export right? Latest commit using tar again for IO looks like it works, and shouldn't have that issue unless I'm mistaken (which is very possible :stuck_out_tongue:)

**Comment by mholt** [comment]
Yeah, and since our current storage APIs don't support streaming, the best we can do is write each file in turn. Maybe in the future we'll be able to stream individual files too.

Thanks for the streaming enhancement! That's going to be much better.

I think we can merge it in this week. I need to do a final review.

File: cmd/commandfuncs.go
Comment (mholt) [inline cmd/commandfuncs.go:547]:
```suggestion
// determineStorage returns the top-level storage module from the given config.
// It may return nil even if no error.
```

File: cmd/commands.go
Comment (mholt) [inline cmd/commands.go:318]:
```suggestion
The - argument refers to stdout and stdin, respectively.
```

File: cmd/commands.go
Comment (mholt) [inline cmd/commands.go:328]: I'm pretty sure autosave.json goes into the config directory (`~/.config/caddy` default on Linux), not storage :thinking:

**Review by mholt — CHANGES_REQUESTED** [review]
Awesome, this is getting close!

See inline comments, plus:

- Maybe we should move the relevant code (other than the `RegisterCommand()`) into a file called `cmd/storage.go` since there's quite a few functions that are needed to make it work, we can keep them together.
- We should add a note to the CLI docs stating that this is experimental and subject to change, maybe: `EXPERIMENTAL: This feature is subject to change or removal.` -- I don't think we'll remove it, but I want to set clear expectations until we iron out any issues later. :)

Thanks for working on this!

File: cmd/commands.go
Comment (francislavoie) [reply cmd/commands.go:318]: I'd put single quotes around it like `'-'` maybe

File: cmd/commands.go
Comment (francislavoie) [reply cmd/commands.go:328]: Correct, config storage is not the same as data storage. Explained here https://caddyserver.com/docs/conventions#file-locations

Comment (Phrynobatrachus) [reply cmd/commands.go]: Thanks! I thought I saw it in exports at one point, but I'm sure I just did something weird myself.

File: cmd/commands.go
Comment (Phrynobatrachus) [reply cmd/commands.go]: `caddy fmt` does not use quotes so I left them out

**Comment by Phrynobatrachus** [comment]
Thanks for the reviews. :)

**Review by mholt — APPROVED** [review]
Awesomesauce, let's try it out!

Thanks for the great contribution.

I'll add you as a collaborator to the repository as we'd be happy to continue collaborating with you, and this way you can make patches or enhancements in a branch and not a fork.

**Comment by mholt** [comment]
I feel like the CI is drunk right now, will try again later

**Comment by Phrynobatrachus** [comment]
Ah I believe that's my bad actually, `errors.Join()` is Go 1.20. Lookin at it.

**Comment by mholt** [comment]
Thank you again!!

**Comment by Phrynobatrachus** [comment]
It's cool to land it! I'll definitely keep an eye out for other contributions, especially anything shaking out from this. Nice working with you both. :point_right:

**Comment by francislavoie** [comment]
Btw, if you'd like to open PRs to fix the `IsTerminal` thing in some of the storage plugins, I'm sure that would be appreciated

**Pull request merged**
